src/cdk/a11y/focus-trap/event-listener-inert-strategy.ts

```diff
@@ -0,0 +1,57 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {FocusTrapInertStrategy} from './focus-trap-inert-strategy';
+import {ConfigurableFocusTrap} from './configurable-focus-trap';
+
+/**
+ * Lightweight FocusTrapInertStrategy that adds a document focus event
+ * listener to redirect focus back inside the FocusTrap.
+ */
+export class EventListenerFocusTrapInertStrategy implements FocusTrapInertStrategy {
+  /** Focus event handler. */
+  private _listener: any;
+
+  /** Adds a document event listener that keeps focus inside the FocusTrap. */
+  preventFocus(focusTrap: ConfigurableFocusTrap): void {
+    this._listener = this._trapFocus.bind(null, focusTrap);
+    focusTrap.ngZone.runOutsideAngular(() => {
+      focusTrap.document.addEventListener('focus', this._listener, true);
+    });
+  }
+
+  /** Removes the event listener added in preventFocus. */
+  allowFocus(focusTrap: ConfigurableFocusTrap): void {
+    focusTrap.document.removeEventListener('focus', this._listener, true);
+    this._listener = null;
+  }
+
+  /**
+   * Refocuses the first element in the FocusTrap if the focus event target was outside
+   * the FocusTrap.
+   */
+  private _trapFocus(focusTrap: ConfigurableFocusTrap, event: FocusEvent) {
+    const target = event.target as HTMLElement;
+    // Don't refocus if target was in an overlay, because the overlay might be associated
+    // with an element inside the FocusTrap, ex. mat-select.
+    if (!focusTrap.element.contains(target) &&
+      target.closest('div.cdk-overlay-pane') === null) {
+        // Some legacy FocusTrap usages have logic that focuses some element on the page
+        // just before FocusTrap is destroyed. For backwards compatibility, wait
+        // to be sure FocusTrap is still enabled before refocusing.
+        setTimeout(() => {
```

Comment:
I think that we might have to wrap this in a `runOutsideAngular` call as well.

Reply:
Okay! Can you help me understand why, before I make the change? Most of my knowledge about running event listeners outside Angular came from this article: https://medium.com/javascript-everyday/adding-event-listeners-outside-of-the-angular-zone-a22f9cfc80eb.

Reply:
It avoids having Angular run another round of change detection since we're not actually changing anything here that would affect a template

Reply:
Sorry, my last question was super vague. Here’s my general understanding so far:
Running code outside Angular means that code won’t trigger Angular’s change detection.
This is safe to do if you know the code won’t change any bindings, because change detection updates the view when bindings change.

We are adding a focus event listener, with a callback function that runs some code wrapped in setTimeout. None of this code changes bindings, so we want to stop change detection from running for any of it. 

Events firing trigger change detection. Javascript timer functions also trigger change detection. Thus two things about our setup could potentially trigger change detection: the focus event firing, and the setTimeout.  

According to the article I linked above, in order for the event firing to not trigger change detection, you need to add the event listener outside Angular. The article doesn’t specify what happens when the callback code includes a timer function that might cause change detection to run again. 

I set up a codepen to experiment with this: https://stackblitz.com/edit/angular-qugmsj. I found that when an event listener is added normally (as in, not outside Angular), crisbeto@ is right that the setTimeout makes change detection run again. So it runs once when the event is fired, and then again due to the setTimeout. In order for change detection to not run at all when an event fires, you must add the event listener outside Angular, as the article stated. At the point you do that, change detection doesn’t run even when the callback has a setTimeout, so there’s no need for another runOutsideAngular in the callback. 

Take a look at the codepen and let me know what you think or if I’ve misunderstood anything! I still don’t fully understand why Angular behaves this way, as it seems somewhat inconsistent.

Reply:
Yeah, you understand it perfectly; since this is an event listener registered outside the zone, the `setTimeout` won't trigger further change detection. I wasn't paying attention to how/where this function was called when I responded to the earlier comment. It might be good to just add a comment that mentions this so that a future person knows why the `runOutsideAngular` isn't necessary here

Reply:
Thank you for confirming! I added a comment to the function.